Adapt project to Airflow v2

saimon46 opened the pull request:
In order to have this plugin compatible with Airflow 2.x these are the changes that I've made:

- I've updated all the classes and did minor changes on the structure. I didn't touch methods and names. Except the poke method in the ClickhouseSqlSensor that had a fix to be compatible with Airflow version 1.10.6, not necessary anymore.
- I've based my changes mostly on the MySqlHook, Sensor and Operator https://github.com/apache/airflow/tree/2.0.1/airflow/providers/mysql
- I've changed the BaseHook dependency with the DbApiHook that has the methods to be overridden.
- I've changed the default signature of the constructors to be compliant with the other operators.
- I've removed the old Plugins class definition not necessary anymore in Airflow v2.
- I've added myself to the contributors.
I've tested the plugin with Airflow v2 and it seems all working fine.
Please review these changes and I'm open to fix possible issues that I've introduced.

Simone

bryzgaloff commented on airflow_clickhouse_plugin/hooks/clickhouse_hook.py:
```diff
@@ -1,24 +1,27 @@
-            clickhouse_conn_id: str = DEFAULT_CONN_ID,
             database: Optional[str] = None,
+            *args,
+            **kwargs
```
Please don't use flexible signature, it might lead to an unexpected behaviour. We are definitely sure this hook uses `clickhouse_conn_id` as an argument, so let's stick with that.

saimon46 replied:
Added it back

bryzgaloff commented on airflow_clickhouse_plugin/hooks/clickhouse_hook.py:
```diff
@@ -1,24 +1,27 @@
-class ClickHouseHook(BaseHook):
-    DEFAULT_CONN_ID = 'clickhouse_default'
+class ClickHouseHook(DbApiHook):
+    conn_name_attr = 'clickhouse_conn_id'
```
Is this used anywhere except `setattr`/`getattr` calls within `DbApiHook` methods?

saimon46 replied:
The DbApiHook needs this variable to work, it's used in all the implementations. If you don't put it the DbApiHook class triggers an exception in the constructor. https://github.com/apache/airflow/blob/a9ac2b040b64de1aa5d9c2b9def33334e36a8d22/airflow/hooks/dbapi.py#L60

bryzgaloff commented on airflow_clickhouse_plugin/hooks/clickhouse_hook.py:
```diff
@@ -1,24 +1,27 @@
     def get_conn(self) -> Client:
-        conn = self.get_connection(self.clickhouse_conn_id)
+        conn = self.get_connection(
+            getattr(self, self.conn_name_attr)
```
Please don't use `getattr`. We have a clear understanding and strict control over structure of hook class attributes, so use explicit attributes names. I mean prefer `self.clickhouse_conn_id` over `getattr(self, self.conn_name_attr)`.

saimon46 replied:
If you prefer so I can put it back ;) I just wanted to be similar to other implementations in other operators.

bryzgaloff commented on airflow_clickhouse_plugin/operators/clickhouse_operator.py:
```diff
@@ -1,18 +1,21 @@
 from airflow_clickhouse_plugin.hooks.clickhouse_hook import ClickHouseHook
 
 
 class ClickHouseOperator(BaseOperator):
```
Is there `DbApiOperator` present in Airflow?

saimon46 replied:
No, the DbApiOperator does not exist.  We have to use the BaseOperator as superclass here.

bryzgaloff commented on airflow_clickhouse_plugin/operators/clickhouse_operator.py:
```diff
@@ -1,18 +1,21 @@
 class ClickHouseOperator(BaseOperator):
     template_fields = ('_sql',)
-    DEFAULT_CONN_ID = ClickHouseHook.DEFAULT_CONN_ID
+    conn_name_attr = 'clickhouse_conn_id'
```
Seems unnecessary

saimon46 replied:
Yes, my mistake, removing it.

bryzgaloff commented:
Hi Simone, huge thanks for your contribution! I am currently reviewing your PR. Will share my feedback soon.

bryzgaloff commented on airflow_clickhouse_plugin/sensors/clickhouse_sql_sensor.py:
```diff
@@ -39,47 +41,4 @@ def _get_hook(self) -> ClickHouseHook:
-                        .format(self.success),
-                )
-        return str(first_cell) not in ('0', '')
+        return super().poke(context)
```
Please add a newline at the end of the file

saimon46 replied:
Done

bryzgaloff commented on setup.py:
```diff
@@ -27,20 +27,10 @@
     install_requires=[
         'clickhouse-driver~=0.1.2',
-        'apache-airflow~=1.10.6',
+        'apache-airflow>=2.0.0',
```
```suggestion
        'apache-airflow~=2.0.1',
```

We may not be sure that 2.1.0 release won't break the plugin.

bryzgaloff commented on setup.py:
```diff
@@ -27,20 +27,10 @@
     python_requires=">=3.6.*",
     install_requires=[
         'clickhouse-driver~=0.1.2',
```
```suggestion
        'clickhouse-driver~=0.1.3',
```

Since you are using DB API, we have to upgrade driver to the first version that supported it: https://github.com/mymarilyn/clickhouse-driver/blob/master/CHANGELOG.md#013---2020-02-21

saimon46 replied:
Updated

bryzgaloff commented on setup.py:
```diff
@@ -27,20 +27,10 @@
     ],
     license=about['__license__'],
     zip_safe=False,
-    entry_points={
```
Could you please share a link to the documentation that states this is not necessary anymore for Airflow v2?

saimon46 replied:
I didn't find any docs that was stating that but this was the error I was getting with the old plugin definition.

```
/usr/local/lib/python3.6/site-packages/airflow_clickhouse_plugin/__init__.py:7: FutureWarning: Registering operators or sensors in plugins is deprecated -- these should be treated like 'plain' python modules, and imported normally in DAGs.

Airflow 2.0 has removed the ability to register these types in plugins. See <http://airflow.apache.org/docs/stable/howto/custom-operator.html>.
  class ClickHouseOperatorPlugin(AirflowPlugin):
/usr/local/lib/python3.6/site-packages/airflow_clickhouse_plugin/__init__.py:17: FutureWarning: Registering operators or sensors in plugins is deprecated -- these should be treated like 'plain' python modules, and imported normally in DAGs.

Airflow 2.0 has removed the ability to register these types in plugins. See <http://airflow.apache.org/docs/stable/howto/custom-operator.html>.
```
Removing the Classes, using directly the Operator and the Hook resolved the issue. If you go here you can see that the plugin is not anymore defined creating a custom Operator: http://airflow.apache.org/docs/stable/howto/custom-operator.html

saimon46 replied:
In this link the change is more clear https://github.com/apache/airflow/blob/2.0.0/UPDATING.md#custom-executors-is-loaded-using-full-import-path

bryzgaloff commented on airflow_clickhouse_plugin/hooks/__init__.py:
```diff
@@ -1 +0,0 @@
-from .clickhouse_hook import ClickHouseHook
```
Which interface do plugins expose with Airflow v2? How should I import the plugin parts now?

saimon46 replied:
I've changed the readme with the new changes on the import.

bryzgaloff commented on README.md:
```diff
@@ -268,22 +266,23 @@ From the root project directory: `python -m unittest discover -s tests`
 [airflow-conn-extra]: https://airflow.apache.org/docs/stable/_api/airflow/models/connection/index.html#airflow.models.connection.Connection.extra
 [ch-driver-connection]: https://clickhouse-driver.readthedocs.io/en/latest/api.html#connection
-[airflow-connection-attrs]: https://airflow.apache.org/docs/1.10.6/_api/airflow/models/index.html?highlight=connection#airflow.models.Connection
-[airflow-conn-dejson]: https://airflow.apache.org/docs/1.10.6/_api/airflow/models/index.html?highlight=connection#airflow.models.Connection.extra_dejson
```
Please use fixed version of docs, so that we commit to support Airflow v2.0.1 and not the dynamic current stable version

bryzgaloff commented on README.md:
```diff
@@ -219,8 +217,8 @@ Important note: don't try to insert values using
-from airflow.sensors.clickhouse_sql_sensor import ClickHouseSqlSensor
-from airflow.operators.clickhouse_operator import ClickHouseOperator
+from airflow_clickhouse_plugin.sensors.clickhouse_sql_sensor import ClickHouseSqlSensor
+from airflow_clickhouse_plugin.operators.clickhouse_operator import ClickHouseOperator
```
Is it a recommended way to import plugins now? Why does former `from airflow.sensors.clickhouse_sql_sensor` not work anymore?

saimon46 replied:
It's a change on Airflow. All the external plugins have to be imported from the root of the module that you want to use.

saimon46 replied:
https://github.com/apache/airflow/blob/2.0.0/UPDATING.md#custom-executors-is-loaded-using-full-import-path here a link that states that change.

bryzgaloff replied:
Looks like you have shared not a correct link, so I will just duplicate it here for the record: https://github.com/apache/airflow/blob/2.0.0/UPDATING.md#adding-operators-and-sensors-via-plugins-is-no-longer-supported

Yes, sounds clear to me, all related to legacy plugin system can be now dropped.

bryzgaloff replied:
Could you please double check and confirm that you have changed it everywhere within the repo? Can be checked by searching for `airflow\.\w+\.clickhouse` pattern.

bryzgaloff reviewed (changes requested):
Looks good in general but please provide some clarifications.

bryzgaloff commented:
Also please verify that tests are passing. Currently all tests are broken. Unfortunately this repo has no CI actions set up, but you may test locally for now: `AIRFLOW_CONN_CLICKHOUSE_DEFAULT='clickhouse://<user>:<password>@<host>:<tcp-port>/<database>?<extra options, e.g. secure=true&ca_certs=/usr/local/share/ca-certificates/Yandex/YandexInternalRootCA.crt>' python -m unittest discover -s tests`

saimon46 commented on README.md:
```diff
@@ -20,9 +20,7 @@ Provides `ClickHouseOperator`, `ClickHouseHook` and `ClickHouseSqlSensor` for
-    1.10.6 since it is the latest version
-    [supported by Google Cloud Composer][cloud-composer-versions]: later
-    versions are expected to work properly but may be not fully tested.
+    2.x. Use the version 0.5.7.post1 to be compatible with Airflow 1.10.6 ~ 1.10.14.
```
Added the fix.

saimon46 commented:
I've made many changes on the tests too but I have 3 tests still failing. Listed them:
- ```
  self.assertTrue(
      ClickHouseSqlSensor(task_id='_', sql='SELECT NULL').poke(None),
  )
   ```
  This come from here: https://github.com/saimon46/airflow-clickhouse-plugin/blob/master/tests/integration/test_sql_sensor.py#L14
- ```
  ======================================================================
  FAIL: test_clickhouse_sql_sensor_poke (unit.test_sql_sensor.ClickHouseSqlSensorTestCase) [[[None]]]
  ----------------------------------------------------------------------
  Traceback (most recent call last):
    File "/Users/sbrundu/Projects/airflow-clickhouse-plugin/tests/unit/test_sql_sensor.py", line 24, in test_clickhouse_sql_sensor_poke
      self.assertEqual(expected_result, op.poke(context=None))
  AssertionError: True != False
  
  ======================================================================
  FAIL: test_clickhouse_sql_sensor_poke (unit.test_sql_sensor.ClickHouseSqlSensorTestCase) [[[0.0]]]
  ----------------------------------------------------------------------
  Traceback (most recent call last):
    File "/Users/sbrundu/Projects/airflow-clickhouse-plugin/tests/unit/test_sql_sensor.py", line 24, in test_clickhouse_sql_sensor_poke
      self.assertEqual(expected_result, op.poke(context=None))
  AssertionError: True != False
  
  ======================================================================
  FAIL: test_clickhouse_sql_sensor_poke (unit.test_sql_sensor.ClickHouseSqlSensorTestCase) [[['0']]]
  ----------------------------------------------------------------------
  Traceback (most recent call last):
    File "/Users/sbrundu/Projects/airflow-clickhouse-plugin/tests/unit/test_sql_sensor.py", line 24, in test_clickhouse_sql_sensor_poke
      self.assertEqual(expected_result, op.poke(context=None))
  AssertionError: False != True
  
  ----------------------------------------------------------------------
  ```
  They come from here: https://github.com/saimon46/airflow-clickhouse-plugin/blob/master/tests/unit/test_sql_sensor.py#L14

Probably this is something that changed on Clickhouse?

saimon46 commented:
Hello everyone, can I have a feedback on this PR? I put all the comments that should help you understand my changes. Please let me know ;)

ne1r0n commented:
Hey @saimon46, initially all test was adopted for the Airflow `1.10.6`.
In the newer versions, return statement in `poke` method of `SqlSensor` was changed from
```return str(first_cell) not in ('0', '')```
to
```return bool(first_cell)```
Therefore, the result in the new version will be different.

@bryzgaloff, please advise what is the best way to deal with it in test cases.

bryzgaloff commented on README.md:
```diff
@@ -268,22 +267,23 @@ From the root project directory: `python -m unittest discover -s tests`
 [ch-driver-execute]: https://clickhouse-driver.readthedocs.io/en/latest/quickstart.html#selecting-data
 [airflow-base-op]: https://airflow.apache.org/docs/stable/_api/airflow/models/baseoperator/index.html
-[airflow-base-hook]: https://airflow.apache.org/docs/stable/_api/airflow/hooks/base_hook/index.html
+[airflow-dbapihook]: https://airflow.apache.org/docs/apache-airflow/stable/_api/airflow/hooks/dbapi/index.html
```
Don't forget to revert here to the base hook.

saimon46 replied:
I've updated the README

bryzgaloff commented on README.md:
```diff
@@ -268,22 +267,23 @@ From the root project directory: `python -m unittest discover -s tests`
-[airflow-conn-dejson]: https://airflow.apache.org/docs/1.10.6/_api/airflow/models/index.html?highlight=connection#airflow.models.Connection.extra_dejson
+[airflow-connection-attrs]: https://airflow.apache.org/docs/apache-airflow/2.0.1/_api/airflow/models/index.html?highlight=connection#airflow.models.Connection
+[airflow-conn-dejson]: https://airflow.apache.org/docs/apache-airflow/2.0.1/_api/airflow/models/index.html?highlight=connection#airflow.models.Connection.extra_dejson
 [airflow-conn-env]: https://airflow.apache.org/docs/stable/howto/connection/index.html#storing-a-connection-in-environment-variables
```
Let's fix version here and in other README files either to 2.0.1

bryzgaloff commented on airflow_clickhouse_plugin/__version__.py:
```diff
@@ -9,8 +9,8 @@
 __url__ = 'https://github.com/whisklabs/airflow-clickhouse-plugin'
 __license__ = 'MIT License'
 __author__ = 'Viktor Taranenko'
-__author_email__ = 'viktor@samsungnext.com'
```
Please add an empty line at the end of the file

bryzgaloff commented on airflow_clickhouse_plugin/operators/clickhouse_operator.py:
```diff
@@ -1,21 +1,24 @@
             database: Optional[str] = None,
-            *args, **kwargs,
+            *args,
+            **kwargs
```
trailing comma

saimon46 replied:
Fixed

bryzgaloff commented on airflow_clickhouse_plugin/sensors/clickhouse_sql_sensor.py:
```diff
@@ -39,47 +41,4 @@ def _get_hook(self) -> ClickHouseHook:
-                        .format(self.success),
-                )
-        return str(first_cell) not in ('0', '')
+        return super().poke(context)
```
No need in overriding it now, simply drop this method

bryzgaloff commented on tests/unit/test_sql_sensor.py:
```diff
@@ -110,8 +110,7 @@ def tearDownClass(cls) -> None:
-        'airflow_clickhouse_plugin.ClickHouseSqlSensor'
-            '._ClickHouseSqlSensor__poke_clickhouse',
+        'airflow_clickhouse_plugin.sensors.clickhouse_sql_sensor.ClickHouseSqlSensor'
     )
```
Not sure if this patch is needed anymore. Does this test fail if we drop that `mock.path` call?

saimon46 replied:
I've removed it and it was failing because the method call in the body of the test. I've removed the assert inside and everything got fixed :)

bryzgaloff commented on tests/unit/test_sql_sensor.py:
```diff
@@ -110,8 +110,7 @@ def tearDownClass(cls) -> None:
     @mock.patch(
-        'airflow_clickhouse_plugin.ClickHouseSqlSensor'
-            '._ClickHouseSqlSensor__poke_clickhouse',
+        'airflow_clickhouse_plugin.sensors.clickhouse_sql_sensor.ClickHouseSqlSensor'
```
trailing comma

bryzgaloff replied:
Also please fit this code line into 80 characters width, e.g.

```suggestion
        'airflow_clickhouse_plugin.sensors.clickhouse_sql_sensor'
            '.ClickHouseSqlSensor',
```

saimon46 replied:
No need anymore since the patch is removed now.

bryzgaloff commented on tests/unit/test_sql_sensor.py:
```diff
@@ -100,7 +100,7 @@ def test_clickhouse_sql_sensor_poke_invalid_success(self):
     def setUpClass(cls):
         cls._get_records_patch = \
-            mock.patch('airflow_clickhouse_plugin.ClickHouseHook.get_records')
+            mock.patch('airflow_clickhouse_plugin.hooks.clickhouse_hook.ClickHouseHook.get_records')
```
```python
        cls._get_records_patch = mock.patch(
            'airflow_clickhouse_plugin.hooks.clickhouse_hook'
                '.ClickHouseHook.get_records',
        )

```

saimon46 replied:
Fixed ;)

bryzgaloff reviewed (changes requested):
@saimon46 thank you for the changes, still some others are required. We are getting closer to the final iteration!

@ne1r0n thank you for your participation.

Regarding `SqlSensor`, let's follow current poke method semantics. So simply update the subtests expected results [here](https://github.com/saimon46/airflow-clickhouse-plugin/blob/master/tests/unit/test_sql_sensor.py#L14-L20) to what current Airflow sensor does. Mean `[[None]]` and `[[0.0]]` -> `False`, `[['0']]` -> `True`.

saimon46 commented on README.md:
```diff
@@ -60,7 +59,7 @@ Supported kwargs of constructor (`__init__` method):
 * `database`: if present, overrides database defined by connection.
 
-Supports all of the methods of the Airflow [BaseHook][airflow-base-hook]
+Supports all of the methods of the Airflow [DbApiHook][airflow-dbapihook]
```
I though that an Deprecation warning would show up importing the clickhouse plugin with the BaseHook import but actually no. I've reverted the change and now the superclass is BaseHook again.

saimon46 commented:
Hello @bryzgaloff and @ne1r0n . I've done the last fixes on the project. I've reverted the DbApiHook superclass in the ClickhouseHook, all good there and there is no Deprecation warning importing directly BaseHook. I've fixed all the tests that now are working and finishing all with success.
Can you please have a last peek of all the changes?

bryzgaloff reviewed (changes requested):
Thank you @saimon46 for all the fixes, now the PR looks awesome. Please make the last change to README.

I can confirm tests are passing. Here are steps to reproduce, hopefully once we will wrap it into a CI pipeline :)

Start container with ClickHouse: `docker exec -it $(docker run --rm -d yandex/clickhouse-server) bash`. Execute within container:
```
apt-get update
apt-get install python3.8 python3-pip git
python3.8 -m pip install apache-airflow==2.0.1 clickhouse-driver~=0.1.3
git clone https://github.com/saimon46/airflow-clickhouse-plugin.git
python3.8 -m unittest discover -s tests
```

saimon46 commented:
I've made the changes you suggested in the README. I didn't add the info to run the tests tho. These is already a recipe on how to do it, not complete as the docker container procedure that you commented above. Do you want to add that too?

bryzgaloff commented:
Thank you @saimon46, no need to include tests recipe into README, this was just for your reference.

Huge thanks for adopting this plugin to Airflow v2!

bryzgaloff commented:
I will try to find some time to publish this to PyPI in the nearest future.

saimon46 commented:
Thank you too @bryzgaloff . I very happy that we managed to merge these changes. We needed this fix on our side and actually we need the package published on PyPI soonish. Can you put an effort on it and publish the new package version? It would be very helpful for us.
Thanks, Simone

bryzgaloff commented:
@saimon46 you can install current master version using `pip install git+https://github.com/whisklabs/airflow-clickhouse-plugin.git`. I plan to release the new version within current week.

bryzgaloff commented:
Hi @saimon46 v0.6.0 is released: https://github.com/whisklabs/airflow-clickhouse-plugin/releases/tag/v0.6.0

Thank you for your participation, enjoy! 😉
